KRFAnalysisPass/KRF.cpp

```diff
@@ -0,0 +1,296 @@
+"rt_sigprocmask",
+"rt_sigqueueinfo",
+"rt_sigsuspend",
+"rt_sigtimedwait",
+"sigaction",
+"sigpending",
+"sigprocmask",
+"sigsuspend",
+"sigtimedwait",
+"sched_get_priority_max",
+"sched_get_priority_min",
+"sched_getaffinity",
+"sched_getattr",
+"sched_getparam",
+"sched_getscheduler",
+"sched_rr_get_interval",
+"sched_setaffinity",
+"sched_setattr",
+"sched_setparam",
+"sched_setscheduler",
+"select",
+"semctl",
+"semget",
+"semop",
+"sendmsg",
+"sendto",
+"setdomainname",
+"setgid",
+"sethostname",
+"setpgid",
+"setpriority",
+"setregid",
+"setresgid",
+"setresuid",
+"setreuid",
+"setsid",
+"setsockopt",
+"settimeofday",
+"setuid",
+"shmat",
+"shmctl",
+"shmdt",
+"shmget",
+"shutdown",
+"sigaltstack",
+"socket",
+"socketpair",
+"stat",
+"statfs",
+"swapoff",
+"swapon",
+"symlink",
+"symlinkat",
+"syncfs",
+"sysfs",
+"syslog",
+"tgkill",
+"time",
+"timer_create",
+"timer_delete",
+"timer_getoverrun",
+"timer_gettime",
+"timer_settime",
+"timerfd_create",
+"timerfd_gettime",
+"timerfd_settime",
+"tkill",
+"truncate",
+"umount",
+"unlink",
+"unlinkat",
+"uselib",
+"ustat",
+"utime",
+"wait4",
+"waitid",
+"write",
+"syscall"
+  };
+					
+  bool runOnModule(Module &M) override {
+    json::OStream* J = (Json) ? new json::OStream(errs()) : NULL;
+    if (Json) {
+      J->objectBegin();
+      J->attributeBegin(M.getName());
+      J->objectBegin();
+    } else {
+      errs() << "KRF: entered module ";
+      errs().write_escaped(M.getName()) << '\n';
+    }
+    for (auto &F : M) {
+      if (Json) {
+	J->attributeBegin((F.hasName()) ? F.getName() : "unname_function");
+	J->arrayBegin();
+      } else {
+	errs() << "  entered function ";
+	errs().write_escaped((F.hasName()) ? F.getName() : "unname_function") << '\n';
+      }
+      for (auto &B : F) {
+	int lookingForErrno = 0;
+	for (auto &I : B) {
+	  if (CallInst* call_inst = dyn_cast<CallInst>(&I)) {
+	    Function * callee = call_inst->getCalledFunction();
+	    if (lookingForErrno && callee && callee->hasName() && callee->getName().equals("__errno_location")) { // If call to errno
+	      for (auto U : call_inst->users()) { // For every instruction that uses that result
+	      	if (LoadInst* load_inst = dyn_cast<LoadInst>(U)) { // Check if its a load
+	      	  for (auto V : load_inst->users()) { // Then for every inst that uses *that* result
+	      	    if (dyn_cast<ICmpInst>(V)) { // Check if its a comparison
+	      	      if (Json) {
+	      		J->attribute("errno_checked", true);
+	      		J->objectEnd();
+	      	      } else {
+	      		errs() << "      errno checked: yes\n";
+	      	      }
+	      	      lookingForErrno = 0;
+	      	      break;
+	      	    }
+	      	  }
+ 	      	}
+	      }
+	      continue;
+	    }
+	    if (lookingForErrno) {
+	      if (Json) {
+		J->attribute("errno_checked", false);
+		J->objectEnd();
+	      }
+	      lookingForErrno = 0;
+	    }			   
+	    if (callee && callee->hasName() && blacklisted_functions.count(callee->getName().str())) {
```

Comment:
You could probably invert this check to flatten the structure a bit: if the callee's name isn't in the list, `continue`.